netket/optimizer/sr/s_jacobian_mat.py

```diff
@@ -0,0 +1,218 @@
+    for all samples σ_i at given values of the parameters p_j
+    Average <O_j> subtracted for each parameter
+    Divided through with sqrt(#samples) to normalise S matrix
+    If scale is not None, columns normalised to unit norm
+    """
+
+    scale: Optional[jnp.ndarray] = None
+    """If not None, contains 2-norm of each column of the gradient matrix,
+    i.e., the sqrt of the diagonal elements of the S matrix
+    """
+
+    x0: Optional[PyTree] = None
+    """Optional initial guess for the iterative solution."""
+
+    @jax.jit
+    def __matmul__(self, vec: Union[PyTree, jnp.ndarray]) -> Union[PyTree, jnp.ndarray]:
+        if not hasattr(vec, "ndim"):
+            vec, unravel = nkjax.tree_ravel(vec)
+        else:
+            unravel = None
+
+        if self.scale is not None:
+            vec = vec * self.scale
+
+        result = jnp.transpose(jnp.conj(jnp.transpose(jnp.conj(self.O @ vec)) @ self.O)) + self.sr.diag_shift * vec
+
+        if self.scale is not None:
+            result = result * self.scale
+
+        if unravel is None:
+            return result
+        else:
+            return unravel(result)
+
+    @jax.jit
+    def _unscaled_matmul(self, vec: jnp.ndarray) -> jnp.ndarray:
+        return jnp.transpose(jnp.conj(jnp.transpose(jnp.conj(S.O @ vec)) @ S.O)) + S.sr.diag_shift * vec
+
+    @jax.jit
+    def solve(self, y: PyTree, x0: Optional[PyTree] = None) -> PyTree:
+        """
+        Solve the linear system x=⟨S⟩⁻¹⟨y⟩ with the chosen iterataive solver.
+
+        Args:
+            y: the vector y in the system above.
+            x0: optional initial guess for the solution.
+
+        Returns:
+            x: the PyTree solving the system.
+            info: optional additional informations provided by the solver. Might be
+                None if there are no additional informations provided.
+        """
+
+        # Ravel input PyTrees, record unravelling function too
+        grad, unravel = nkjax.tree_ravel(grad)
+        
+        if x0 is None:
+            x0 = self.x0
+        if x0 is not None:
+            x0, _ = nkjax.tree_ravel(x0)
+            if self.scale is not None:
+                x0 = x0 * self.scale
+
+        if self.scale is not None:
+            grad = grad / self.scale
+
+        solve_fun = self.sr.solve_fun()
+        out, info = solve_fun(self._unscaled_matmul, grad, x0=x0)
+
+        if self.scale is not None:
+            out = out / self.scale
+
+        return unravel(out), info
+
+    @jax.jit
+    def to_dense(self) -> jnp.ndarray:
+        """
+        Convert the lazy matrix representation to a dense matrix representation.
+
+        Returns:
+            A dense matrix representation of this S matrix.
+        """
+        if scale is None:
+            O = self.O
+            diag = jnp.eye(self.O.shape[1])
+        else:
+            O = self.O * self.scale[jnp.newaxis,:]
+            diag = jnp.diag(jnp.square(self.scale))
```

Comment:
I prefer more clear code when possible and hate stuff like  `.square`.
In any case they compile down to the same thing
```python
>>> a=jax.numpy.ones(3)
>>> def f(x):
...   return jax.numpy.square(x)
...
>>> def g(x):
...   return x**2
...
>>> jax.make_jaxpr(f)(a)
{ lambda  ; a.
  let b = integer_pow[ y=2 ] a
  in (b,) }
>>> jax.make_jaxpr(g)(a)
{ lambda  ; a.
  let b = integer_pow[ y=2 ] a
  in (b,) }
```

```suggestion
            diag = jnp.diag(self.scale**2)
```

Reply:
Cool. I was using `jnp.*` forms everywhere because I wasn't sure something like `x**2` or member functions would be evaluated on the GPU. But this is not a concern then?

Reply:
Jax first transforms code to IR (intermediate representation), what you see above, and then compiles it to a backend.
CPU/TPU is completely orthogonal concept

Reply:
What I meant is that if the input is (say) a `np.ndarray`, then `x**2` would default to the numpy routine that would run on the CPU even if a GPU is available. But of course, we made sure these objects are `jnp.ndarray`s, so this isn't a concern...

Reply:
Ah I see.
It makes sense while prototyping but here we're so deep in the stack (and we're in a jitted block) so stuff is surely where you need it to be.